src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,160 @@ def use_quotas(
             indicates how much quota should actually be consumed.
         """
         pass
+
+
+class RedisCardinalityLimiter(CardinalityLimiter):
+    def __init__(self, cluster: str = "default", cluster_shard_factor: int = 3) -> None:
+        self.client = redis.redis_clusters.get(cluster)
+        self.cluster_shard_factor = cluster_shard_factor
+        super().__init__()
+
+    @staticmethod
+    def _get_timeseries_key(request: RequestedQuota, hash: Hash) -> str:
+        return f"cardinality-counter-{request.prefix}-{hash}"
+
+    def _get_read_sets_keys(
+        self, request: RequestedQuota, quota: Quota, timestamp: Timestamp
+    ) -> Sequence[str]:
+        oldest_time_bucket = list(quota.iter_window(timestamp))[-1]
+        return [
+            f"cardinality-sets-{request.prefix}-{shard}-{oldest_time_bucket}"
+            for shard in range(self.cluster_shard_factor)
+        ]
+
+    def _get_write_sets_keys(
+        self, request: RequestedQuota, quota: Quota, timestamp: Timestamp, hash: Hash
+    ) -> Sequence[str]:
+        shard = hash % self.cluster_shard_factor
+        return [
+            f"cardinality-sets-{request.prefix}-{shard}-{time_bucket}"
+            for time_bucket in quota.iter_window(timestamp)
+        ]
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[Timestamp] = None
+    ) -> Tuple[Timestamp, Sequence[GrantedQuota]]:
+        if timestamp is None:
+            timestamp = int(time.time())
+        else:
+            timestamp = int(timestamp)
+
+        unit_keys_to_get: List[str] = []
+        set_keys_to_count: List[str] = []
+
+        for request in requests:
+            if request.quotas:
+                for hash in request.unit_hashes:
+                    unit_keys_to_get.append(self._get_timeseries_key(request, hash))
+
+            for quota in request.quotas:
+                set_keys_to_count.extend(self._get_read_sets_keys(request, quota, timestamp))
+
+        if not unit_keys_to_get and not set_keys_to_count:
```

Comment:
When can this happen ?

Reply:
this is to avoid hitting redis when there are no quotas to enforce.